### DISCUSSION TITLE
Unnecessary "observe" warning

### DISCUSSION THREAD

**Discussion by nmalevich** [pointfreeco/swift-composable-architecture#3006] [Q&A]
### Description

Warning "An "observe" was called from another "observe" closure, which can lead to over-observation and unintended side effects." was added in version 1.9.3 and become at unnecessary cases
I attach example project where you can see that bug
[BadObserveExample.zip](https://github.com/pointfreeco/swift-composable-architecture/files/15078857/BadObserveExample.zip)

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Warning doesn't appear

### Actual behavior

Warning appears

### Steps to reproduce

Just run app and wait. Screens will navigate by themselves

### The Composable Architecture version information

1.9.3

### Destination operating system

iOS 17.4.1

### Xcode version information

15.3

### Swift Compiler version information

```shell
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
```

**Comment by stephencelis** [comment]
@nmalevich We recently added this runtime warning because nested observes can have unintended consequences. In this case, a `viewDidLoad` is evaluated in the same stack as the presenting view controller's `observe`, which means all child state evaluated in the child controller's `observe` will be observed by the parent.

To work around the problem, you can move `observe` into `viewWillAppear` instead, and manage the lifetime of the observation token, or you can dispatch the observation a tick in `viewDidLoad`:

```swift
DispatchQueue.main.async {
  observe { [weak self] in
    …
  }
}
```

We're still deciding if this warning is too heavy-handed, but we don't consider it a bug at the moment. Because of this I'm going to convert this to a discussion.

**Reply by nmalevich** [reply]
@stephencelis Hi!
Thank you for your fast answer! But I don't understand exactly why it happens. You sad something about same stack, what did you mean? And why it happens only after presenting the third screen, not after the second?

Yeah, I have seen topic about cancellation observing on viewWillDisappear, but we can't do it with navigation, because we wouldn't be able to dismiss presented controller

I will be really grateful for your explanation. Thanks

**Reply by langrisser1981** [reply]
If I use a custom container to display view controllers, for example addChild
the same warning appears and the observe is triggered repeatedly. 
I think it should be the "unintended consequences" mentioned earlier.

using the same example above, for LaunchViewController, instead of using pushViewController, I use the following method to add introViewController:

```swift
// navigationController?.pushViewController(
//     introController!,
//     animated: true
// )
addChild(introController!)
view.addSubview(introController!.view)
introController?.didMove(toParent: self)
```

also add a print statement inside the observe block to monitor if observe is triggered:
```swift
print("launchVC observe invoked")
```

then run, it will also triggers the warning. 
At the same time, I find that as subsequent view controllers are added, the observe in launchVC keeps being triggered:

```
launchVC observe invoked
launchVC observe invoked
launchVC observe invoked
An "observe" was called from another "observe" closure, which can lead to over-observation and unintended side effects.
Avoid nested closures by moving child observation into their own lifecycle methods.
introVC observe invoked
introVC observe invoked
An "observe" was called from another "observe" closure, which can lead to over-observation and unintended side effects.
Avoid nested closures by moving child observation into their own lifecycle methods.
Optional("")
launchVC observe invoked
```

try using _printChanges to see what happened, I find that it's caused by changes to LaunchFeature.State.destination.
it seems to only happen in the custom containers mentioned earlier or when use present, but not with pushViewController.

I'm not sure if this is normal behavior.
